### DISCUSSION TITLE
tzdata

### DISCUSSION THREAD

**Discussion by edwardp3** [lxqt/lxqt#2107] [Q&A]
Hi everyone. I have a few virtual machines runnning on Debian, using virt-manager to create them. The tzdata package with PCLinuxOS was updated recently and this morning, noticed the clock in the LXQt panel for it, was off by one hour, it was displaying one hour earlier than the correct time. Going into the LXQt Date & Time Settings, that window is showing the correct time. Would this actually be an issue with tzdata or something else? The panel clock was accurate before this update.

Thanks in advance.

![clock-pclos](https://user-images.githubusercontent.com/73129450/139698503-9294b6f0-50cf-46e4-a0fc-720ba279f41e.png)
![clock-pclos-2](https://user-images.githubusercontent.com/73129450/139698517-ae68fecb-f840-4536-aa6a-b7c20bc4cd14.png)

**Comment by edwardp3** [comment]
I just booted up the other three VM's I'm running, Fedora, Lubuntu and Mageia. Lubuntu and Fedora are still displaying the correct time. Mageia, which also displayed the correct time earlier today, is now also displaying the incorrect time, off by the same one hour as PCLinuxOS.

Any ideas what's going on??

**Reply by stefonarch** [reply]
This isn't related to legal/summer time? The change (-1 hour) was Sunday morning at 3:00 in Europe.

**Reply by tsujan** [reply]
A problem in tzdata package is the most probable cause that comes to my mind.

I think you might get better answers at Mageia's/PCLinuxOS's forums. However, another Mageia/PCLinuxOS user may also find this page and reply.

**Comment by edwardp3** [comment]
Hi. I'm in the U.S. and the states that observe Standard Time, will change the clocks (-1 hour) at 2:00 AM local time Sunday November 8. The UTC offset then becomes -05:00. 

If it _is_ somehow related, it shouldn't change here until next week. 

I checked my other VM's that are installed. Fedora and Lubuntu continue to display the correct time. The time in the Mageia (France) panel was fine earlier this morning, but it is also now off by one-hour. PCLinuxOS is based in the U.S. and I installed the same PCLOS image in a new VM without updating anything and that is also displaying the incorrect time, like its original VM, off by the same -1 hour. 

Why would this affect Mageia and PCLOS, but not Fedora and Lubuntu? This doesn't make sense.

**Reply by tsujan** [reply]
> Why would this affect Mageia and PCLOS, but not Fedora and Lubuntu?

→ https://github.com/lxqt/lxqt/discussions/2107#discussioncomment-1571276

**Reply by edwardp3** [reply]
Hi Tsujan - None of this is making sense, but it would seem that the issue is with tzdata.

I have posted the discovery to the PCLOS forum. I am also registered on Mageia's forum, but because it was my first post, it must be approved by a moderator before it will show up.

**Reply by tsujan** [reply]
> None of this is making sense

Some computer problems don't make sense but only at first. Since you said there's no problem in Fedora and Ubuntu, if they have the same version of tzdata as Mageia and PCLinuxOS do, the problem *may be* in packaging; if not, the tzdata version in Mageia/PCLinuxOS *may have* a bug.

To make sure, you could try a stand-alone clock app (I don't remember the name of any but there should be many — I have a Cairo+LUA based analog clock inside Conky).

EDIT: Just use `date` in a terminal.

**Reply by tsujan** [reply]
> Going into the LXQt Date & Time Settings, that window is showing the correct time

Sorry, I didn't pay enough attention to your first comment. LXQt Date & Time Setting serves as a stand-alone clock here, although trying a third clock would be helpful.

If they show the correct time but lxqt-panel's clock doesn't, the problem couldn't be in tzdata. Then, everything I said about tzdata can be repeated about lxqt-panel.

EDIT:

Simply entering `date` in a terminal is enough.

**Reply by edwardp3** [reply]
Entering `date` in the terminal, displayed the correct time.

Both Mageia and PCLinuxOS' package is `timezone` and the versions are different. PCLinuxOS has `2021e`, while Mageia has `2021a`.

On the Mageia forum, someone replied with a link to a bug report (https://bugs.mageia.org/show_bug.cgi?id=29591) mentioning a Qt function, which mentions Plasma and LXQt.

**Reply by tsujan** [reply]
I have Manjaro with the latest git LXQt and latest stable versions of everything, Qt5 included (patched by KDE devs). I have "New York" among the time-zones of lxqt-panel's World Clock. it shows the correct time (02:20:56 pm at this moment).

EDIT:
My tzdata is `2021d` but `2021e` will come here tonight or tomorrow.

**Reply by edwardp3** [reply]
I found Mageia did not have `ntp` and `ntpdate` installed, I installed them, then had to manually start ntpdate.service. `ntpd -q` listed the log up to the point it sync'd the time, but in the end the panel time was still off by one hour.

Unless it _is_ that Qt function in the Mageia bug report, I don't know...

**Reply by tsujan** [reply]
> Unless it is that Qt function in the Mageia bug report, I don't know...

The Qt5 version is `5.15.2+kde+r254` here (`kde+r254` refers to the latest patch that's backported from Qt itself by KDE because Qt doesn't have an open-source release for later versions of Qt5.15).

**Reply by edwardp3** [reply]
Mageia has 0.16 of LXQt, PCLinuxOS uses 0.17. About LXQt in all four of the OS' (plus Debian as the host) shows they're all built with Qt 5.15.2. 

Installing `ntp` and `ntpdate` in Mageia didn't fix the problem, so they were uninstalled. It's not the virtual environment, because one might think all four would have this problem. But there's clearly something different with Fedora and Lubuntu, where this isn't an issue with those.

**Reply by jopamo** [reply]
First off, thanks tsujan for your work. :) 

I am chasing the same issue. lxqt-panel worldclock is slow one hour. System clock is correct. I am using a customized version of Gentoo. I am using git versions of all lxqt packages and qt 5.15 from kde as well. It has been 22 days since I had updated tzdb and this just started a few days ago, so tzdb is not the likely cause IMHO. I will update here if I have luck figuring it out.

**Reply by tsujan** [reply]
It's still OK here, with London, Berlin and New York. If it happens to me, I think I could easily find the culprit by checking the update log (`/var/log/pacman.log`).

**Reply by jopamo** [reply]
I only had updated related git packages around that timeframe and looking at logs, nothing is related to time, so it is not apparent there.

edwardp3, can you easily dump your installed packages list to a pastebin from both of those distros, so I can see what the common denominators are. I cannot find an online package browser. 

If I manually set system date to October 31 2021 or less, the lxqt-panel time is correct. If I manually set system date to Nov 1 2021 to Nov 6 2021, lxqt-panel clock is slow 1 hour. Starting on Nov 7 2021 it is right again. Jumping ahead or back years, I cannot reproduce. It is only Nov 1-Nov 6 2021 it seems. So the trigger was the date itself not a recent installed package.

**Reply by stefonarch** [reply]
Just arrived tzdata in debian, no issue. 

> So the trigger was the date itself not a recent installed package



> Hi. I'm in the U.S. and the states that observe Standard Time, will change the clocks (-1 hour) at 2:00 AM local time Sunday November 8.

So it looks like happening in the gap betweeb change from summertime Europe/US (change in Europa was 31 morning)?

**Reply by jopamo** [reply]
I forgot to try more timezones settings. Looks like when system time zone is in Europe ( I tried Berlin and Dublin ), time is correct. UTC is also correct. When system timezone set to any America timezone (Los Angeles, Denver, Chicago, New York), lxqt-panel shows an hour slow. 

I went ahead and installed Mageia. Going through installed packages list now.

looks like stefonarch is right.
 
date -s "07 NOV 2021 01:00:00" shows 12am
date -s "07 NOV 2021 02:00:00" shows proper 2am

**Reply by edwardp3** [reply]
> edwardp3, can you easily dump your installed packages list to a pastebin from both of those distros, so I can see what the common denominators are. I cannot find an online package browser.

As to the above, no. I have since removed Maegia from the list of VM's I am running. PCLinuxOS uses a package 'timezone', the current version is 2021e.

I also ran PCLOS' LXQt image from late last year (2020.10) as is, without updating any packages into RAM, to find the same, time off by one hour. So whatever is causing this, it seems to have been in place for at least a year, if not longer.

**Reply by jopamo** [reply]
Here is the command to list installed packages for rpm based distros fyi for next time:
`rpm -qa`

I had no common denominators with Maegia. checking pclinux is probably a waste of time.

rolling back to tzdb-2020a shows the correct time, everything after it up to 2021e shows slow one hour. 

I decided to repackage 2021e based on the simpler arch pkgbuild and it works! 

Thanks for your help guys, sorry for the trouble, this was a weird one.

**Comment by tsujan** [comment]
This may also be related to the absence of systemd. Apparently, only systems without systemd are affected. If that's the case, please open a bug report!

**Reply by edwardp3** [reply]
As I switched to Lubuntu from Debian, I am no longer running the VM's. I do not know about PCLinuxOS, but Mageia (which also had this problem) uses systemd.

**Reply by tsujan** [reply]
> I do not know about PCLinuxOS, but Mageia (which also had this problem) uses systemd.

OK. So, please ignore my comment. After all, what Panel's clock shows is based on Qt methods, not systemd.

**Comment by jopamo** [comment]
**This post is dynamic to be as accurate as possible**

This is easily reproducible, I do not think any package versions matter much besides 'tzdata' and it only needs to be >2020a for the new upstream default of 'slim'. After 2020a, you can force slim by running `zic -b slim ...` to rebuild your tzdata or you can just make sure it is not being disabled in your packaging. Most distros disable the new behavior in one of two ways. You can either recompile the timezone database package, appending to cpp-flags `-DZIC_BLOAT_DEFAULT=\"fat\"` or at runtime, use the option `zic -b fat ...` to (re)build timezone data. Make sure you have 'slim' files by comparing to hex dump below and then as superuser, `timedatectl set-timezone America/New_York && date -s "05 NOV 2021 11:00:00"`. lxqt-panel will show 10:00am while everywhere else shows proper 11am. 

Here's some background info:

**upstream, the glib post is more clear in explaining the differences**
https://mm.icann.org/pipermail/tz/2020-June/029125.html
https://gitlab.gnome.org/GNOME/glib/-/issues/2129

**gentoo (optional, default of fat)**
https://bugs.gentoo.org/747538
https://gitweb.gentoo.org/repo/gentoo.git/tree/sys-libs/timezone-data/timezone-data-2021e.ebuild

**arch (fat)**
https://github.com/archlinux/svntogit-packages/blob/packages/tzdata/trunk/PKGBUILD
https://bugs.archlinux.org/task/68150
https://github.com/archlinux/svntogit-packages/commit/7d8fb81c3f5b3376c5c81fe3d8e287d5bab04562

**alpine (fat)**
https://github.com/alpinelinux/aports/commit/607c839c1595e965cd3814124bdbd6d92a4cb862

**void (fat)**
https://github.com/void-linux/void-packages/commit/a5b124327442d90e4842b71a0599a5e3f48ab540

**sample hex of 'America/New_York'**
`xxd /usr/share/zoneinfo/America/New_York`
[fat TZif sample file](http://ix.io/3NNc)
[slim TZif sample file](http://ix.io/3NNd)

**Ubuntu, Debian, and Fedora are all confirmed to be using 'fat' by comparing to above.**
ubuntu ( ubuntu-21.10-desktop-amd64.iso - 2021a-2ubuntu1 )
debian ( debian-live-11.1.0-amd64-lxqt.iso - 2021a-1+deb11u1 )
fedora ( Fedora-Workstation-Live-x86_64-35-1.2.iso - tzdata-2021b-1.fc35.noarch )

**tldr:** A number of apps including QT5 malfunctioned when using the somewhat new default (starting with 2020b) of -DZIC_BLOAT_DEFAULT=\"slim\" in the Timezone Database package, so most linux distros have reverted this change one way or another. 

**confirmed fix for this particular QT5 issue**
https://bugreports.qt.io/browse/QTBUG-90553
https://github.com/qt/qtbase/commit/38a72af4fd311219779f89b2b78cd56d879f61db

**kde 5.15 repo PR**
https://invent.kde.org/qt/qt/qtbase/-/merge_requests/79

**thanks:**
@edwardp3 @tsujan @stefonarch 
and the people at Mageia for feedback.

**Reply by jopamo** [reply]
https://invent.kde.org/qt/qt/qtbase/-/commit/d44d93f64d6b67513cd4e7a80ff331d7e5275389
